### PR TITLE
#1174 add ChText (CommitHash from a text source) - accessing HashOfTag from file

### PR DESCRIPTION
**Pull request by volodya-lombrozo**
Original issue: #1174.

1. Added `ChText` implementation
2. Added `offlineHashFile` flag in `PullMojo`
3. Added tests to check both classes.
4. Added several puzzles

The PR isn't final, we still have to implement `-DofflineHash=*.*.*:a1b2c3d` - it's the next step.

### CODE REVIEW

**Review by Graur — CHANGES_REQUESTED** [review]
@volodya-lombrozo Good one, just a few comments below:

File: eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public String value() {
+        try {
+            return Arrays.stream(new Unchecked<>(this.source).value().split("\n"))
+                .filter(s -> s.contains(this.tag))
+                .findAny()
+                .orElseThrow(NotFound::new)
+                .split("\\s+")[0];
+        } catch (final NotFound | UncheckedIOException ex) {
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java]: @Graur not actual anymore.

File: eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    ChText(final Scalar<String> source, final String tag) {
+        this.source = source;
+        this.tag = tag;
+    }
+
+    @Override
+    public String value() {
+        try {
+            return Arrays.stream(new Unchecked<>(this.source).value().split("\n"))
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java]: @Graur done.

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 3]
```diff
@@ -54,14 +54,19 @@ public final class PullMojo extends SafeMojo {
 
     /**
      * The Git hash to pull objects from, in objectionary.
+     *
      * @since 0.21.0
+     * @todo #1174:90min The wrong naming. It isn't a `hash` - it's a `tag`.
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:59]: @Graur Here is [downloaded file](https://home.objectionary.com/tags.txt):
```
434868a411b9741fdd4f8a38a5c576e8733345c9	gh-pages
a56ee36807fae45ab3ef497451b1066bdd7704c5	master
db0579b48dd717ac015b7af8515f8a56ee6fd521	renovate/actions-checkout-3.x
5fe5ad8d21dbe418038fa4c86e096fb037f290a9	0.23.15
15c85d7f8cffe15b0deba96e90bdac98a76293bb	0.23.17
4b19944d86058e3c81e558340a3a13bc335a2b48	0.23.19
0aa6875c40d099c3f670e93d4134b629566c5643	0.25.0
ff32e9ff70c2b3be75982757f4b0607dc37b258a	0.25.5
e0b783692ef749bb184244acb2401f551388a328	0.26.0
cc554ab82909eebbfdacd8a840f9cf42a99b64cf	0.27.0
00b60c7b2112cbad4e37ba96b162469a0e75f6df	0.27.2
6a70071580e95aeac104b2e48293d3dfe0669973	0.28.0
0c15066a2026cec69d613b709a301f1573f138ec	0.28.1
9b883935257bd59d1ba36240f7e213d4890df7ca	0.28.10
54d83d4b1d28075ee623d58fd742eaa529febd3d	0.28.2
6c6269d1f9a1c81ffe641538f119fe4e12706cb3	0.28.4
9c9352890b5d30e1b89c9147e7c95a90c9b8709f	0.28.5
17f89293e5ae6115e9a0234b754b22918c11c602	0.28.6
5f82cc1edffad67bf4ba816610191403eb18af5d	0.28.7
be83d9adda4b7c9e670e625fe951c80f3ead4177	0.28.9
```

As I understand the right column is a tag column, the left one - for hashes.
In the field declaration you can see the next initialisation:
```java
   @Parameter(property = "eo.hash", required = true, defaultValue = "master")
    private String hash = "master";
```
It means that the default value is `master`. I don't think it's a hash.
Moreover, further we work with it as with tag. 
```java
new ChRemote(this.hash);
...
ChRemote(final String tag) {
    this.tag = tag;
}
```
We definitely have some confusion which should be fixed.

[the remark this answers is not in the document]
Comment (Graur) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:59]: @volodya-lombrozo Agree, I didn't see these changes with tags and hashes

**Review by Graur — APPROVED** [review]
@volodya-lombrozo Looks good to me
@mximp @yegor256 Do you have any comments here?

**Comment by volodya-lombrozo** [comment]
@yegor256 reminder

File: eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                                }
+                            )
+                        ),
+                        "\\s+"
+                    ),
+                    () -> {
+                        throw new NotFound();
+                    }
+                )
+            )).value();
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java:100]: @volodya-lombrozo check this: https://www.yegor256.com/2014/10/23/paired-brackets-notation.html

Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/ChText.java]: @yegor256 solved.

**Comment by volodya-lombrozo** [comment]
@yegor256 the brackets problem is solved. Please, merge if you don't have any objections

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**